File: 0329-Longest Increasing Path in a Matrix/0329-Longest Increasing Path in a Matrix.cpp
class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int m = matrix.size();
        if (m == 0)
        {
            return 0;
        }
        int n = matrix[0].size();
        vector<vector<int>> cache(m, vector<int>(n));
        int maxLen = 0;
        for (int i = 0; i < m; ++i)
        {
            for (int j = 0; j < n; ++j)
            {
                maxLen = max(maxLen, dfs(matrix, i, j, cache));
            }
        }
        
        return maxLen;
    }
    
private:
    int dfs(vector<vector<int>>& matrix, int x, int y, vector<vector<int>>& cache)
    {
        if (cache[x][y])
        {
            return cache[x][y];
        }
        
        int m = matrix.size();
        int n = matrix[0].size();
        
        vector<int> dx = {-1, 1, 0, 0};
        vector<int> dy = {0, 0, -1, 1};
        int maxLen = 1;
        for (int k = 0; k < dx.size(); ++k)
        {
            int nx = x + dx[k];
            int ny = y + dy[k];
            if (nx >= 0 && nx < m && ny >= 0 && ny < n && matrix[nx][ny] > matrix[x][y])
            {
                maxLen = max(maxLen, 1 + dfs(matrix, nx, ny, cache));
            }
        }
        
        return cache[x][y] = maxLen;
    }
};

File: 0301-0400/0329-Longest Increasing Path in a Matrix/0329-Longest Increasing Path in a Matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        if not matrix:
            return 0
        m, n = len(matrix), len(matrix[0])
        cache = [[0] * n for _ in range(m)]
        maxLen = 0
        def dfs(r, c):
            if cache[r][c] == 0:
                ml = 1
                for nr, nc in (r-1, c), (r+1, c), (r, c-1), (r, c+1):
                    if 0 <= nr < m and 0 <= nc < n and matrix[nr][nc] > matrix[r][c]:
                        ml = max(ml, 1 + dfs(nr, nc))                        
                cache[r][c] = ml
            return cache[r][c]
        
        for i in range(m):
            for j in range(n):
                maxLen = max(maxLen, dfs(i, j))
        return maxLen
